### PR TITLE
Update poetry to version 1.2.0

### PR DESCRIPTION
**Pull request by danyeaw**
This PR updates Poetry to version 1.2.0. In order to do that, it replaces the legacy way that poetry group dev and non-dev dependencies and creates new groups. An overview of this is in the [poetry 1.2.0 announcement](https://python-poetry.org/blog/announcing-poetry-1.2.0/)

### PR Checklist
Please check if your PR fulfills the following requirements:

- [X] I have read, and I am following the [Contributor guide](https://github.com/gaphor/gaphor/blob/main/CONTRIBUTING.md)
- [X] I have read, and I understand the [Code of Conduct](https://github.com/gaphor/gaphor/blob/main/CODE_OF_CONDUCT.md)

### PR Type
What kind of change does this PR introduce?

- [ ] Bug fix
- [ ] Feature
- [X] Chore (refactoring, formatting, local variables, other cleanup)
- [ ] Documentation content changes

### What is the current behavior?

Issue Number: N/A

### Does this PR introduce a breaking change?
- [ ] Yes
- [X] No

### CODE REVIEW

**Comment by amolenaar** [comment]
The group functionality looks like a nice improvement over the current approach :+1:

**Comment by danyeaw** [comment]
image
PyInstaller isn't packaging everything we need for some reason.

File: gaphor/ui/selftest.py [hunk 1]
```diff
@@ -1,14 +1,33 @@
...
+            env_path,
+        ),
+        None,
+    ):
+        with os.add_dll_directory(first_gtk_path):
+            import gi
+
+else:
+    import gi
+
```
Comment (amolenaar) [inline gaphor/ui/selftest.py:28]: There should be no need to put it here, if it's already defined in the `__init__.py` of the same module.

File: gaphor/ui/__init__.py [hunk 2]
```diff
@@ -1,13 +1,31 @@
...
+            and Path.is_file(Path(path) / "girepository-1.0-1.dll"),
+            env_path,
+        ),
+        None,
+    ):
+        with os.add_dll_directory(first_gtk_path):
+            import gi
+
+else:
+    import gi
```
Comment (amolenaar) [inline gaphor/ui/__init__.py:28]: Why do we need this? Can't we use the old way instead?

**Review by amolenaar — CHANGES_REQUESTED** [review]
I think this PR is mixing 2 things: updates for Poetry 1.2 and a new way to install stuff on Windows.

Is there a reason to combine these? I like the improvements for Poetry, since it removes a lot of "extras" from the distribution.

I'd rather not have the win32 `add_dll_directory` workaround.

**Comment by danyeaw** [comment]
Unfortunately Poetry 1.2.0 will update installed packages even if I already have the correct version installed. So our previous method of installing the wheels from gvsbuild and then installing the other packages no longer works.

**Comment by amolenaar** [comment]
The win32 patch can also be applied to 'gaphor/__init__.py', since Gaphas also requires this. Or a special module, similar to macos shim module.

**Comment by danyeaw** [comment]
@amolenaar Ok, I refactored the logic and created a windowsshim module. Unfortunately, I can't get PyInstaller to work without https://gitlab.gnome.org/GNOME/pygobject/-/merge_requests/211 changes to PyGObject directly.

**Comment by amolenaar** [comment]
This looks good. I think we do not need the same code in `selftest.py`. It's unfortunate we can't see why the windows self-test job is failing :disappointed: .

**Comment by amolenaar** [comment]
Moving the windows fix up doesn't resolve the issue. It's like the import is already handled in the pyimod2- thingy 🤔

**Comment by amolenaar** [comment]
I created 0948b674bcba1478c207ffc862b39fddd15cdb06 (branch poetry-1.2-test). This fixes the windows distribution.

Do we still need the workaround in our code then?

**Comment by danyeaw** [comment]
Nice one @amolenaar, I didn't consider overriding the installed wheel after the poetry run.

pycairo doesn't need the workaround any longer, so we only need to override PyGObject, I updated the PR with just that one.

I think the workaround is still good, otherwise every time we run `poetry install` in Windows we need to then run `pip install --force-reinstall C:\gtk-build\build\x64\release\pygobject\*.whl` which is pretty annoying.

We should also squash and merge this one, somehow the git history got crazy ugly when rebasing - I must have run a git pull.

**Review by amolenaar — APPROVED** [review]

**Comment by amolenaar** [comment]
> We should also squash and merge this one, somehow the git history got crazy ugly when rebasing - I must have run a git pull.

I fixed the git history and merged it. A nice improvement :+1:.

**Pull request merged**